src/solver/multilabel_margin_loss/multilabel_margin_loss.cpp

```diff
@@ -0,0 +1,209 @@
+        {"MIOPEN_USE_FP16", static_cast<int>(dtype == miopenHalf)},
+        {"MIOPEN_USE_FP32", static_cast<int>(dtype == miopenFloat)},
+        {"MIOPEN_USE_BFP16", static_cast<int>(dtype == miopenBFloat16)},
+        {"IN_OUT_TYPE", i_dtype == "bfloat16" ? "ushort" : i_dtype},
+        {"TARGET_TYPE", t_dtype},
+    };
+
+    kernel.comp_options = build_params.GenerateFor(kbp::HIP{});
+
+    size_t xlocalsize = LOCAL_SIZE;
+    size_t xgridsize  = AlignUp(N, xlocalsize);
+    size_t ylocalsize = 1;
+    size_t ygridsize  = 1;
+    size_t zlocalsize = 1;
+    size_t zgridsize  = 1;
+    kernel.l_wk.push_back(xlocalsize);
+    kernel.l_wk.push_back(ylocalsize);
+    kernel.l_wk.push_back(zlocalsize);
+
+    kernel.g_wk.push_back(xgridsize);
+    kernel.g_wk.push_back(ygridsize);
+    kernel.g_wk.push_back(zgridsize);
+
+    result.construction_params.push_back(kernel);
+
+    // Construct reduce kernel params
+    auto _size = size;
+    do
+    {
+        auto reduce_kernel        = KernelInfo{};
+        reduce_kernel.kernel_file = "MIOpenLossSum.cpp";
+        reduce_kernel.kernel_name = "LossSum";
+
+        reduce_kernel.comp_options = build_params.GenerateFor(kbp::HIP{});
+
+        reduce_kernel.l_wk.push_back(xlocalsize);
+        reduce_kernel.l_wk.push_back(ylocalsize);
+        reduce_kernel.l_wk.push_back(zlocalsize);
+
+        reduce_kernel.g_wk.push_back(AlignUp(N, LOCAL_SIZE_REDUCE));
+        reduce_kernel.g_wk.push_back(ygridsize);
+        reduce_kernel.g_wk.push_back(zgridsize);
+
+        result.construction_params.push_back(reduce_kernel);
+        _size = AlignUp(_size, LOCAL_SIZE_REDUCE) / LOCAL_SIZE_REDUCE;
+    } while(_size > 1);
+
+    result.invoker_factory = [](const std::vector<Kernel>& kernels) {
+        return [=](const Handle& handle_, const AnyInvokeParams& raw_params) {
+            decltype(auto) params =
+                raw_params.CastTo<miopen::multilabel_margin_loss::InvokeParams>();
+            /* Phase 1: Calc loss for each element. */
+            auto idims = params.iDesc->GetLengths();
+            auto tdims = params.tDesc->GetLengths();
+
+            auto istrides   = params.iDesc->GetStrides();
+            auto tstrides   = params.tDesc->GetStrides();
+            auto N          = idims[0];
+            auto odtypeSize = get_data_size(params.oDesc->GetType());
+
+            int64_t ws_offset =
+                ((N) + ((N + LOCAL_SIZE_REDUCE - 1) / LOCAL_SIZE_REDUCE)) * odtypeSize;
+            {
+                decltype(auto) kernel = handle_.Run(kernels[0]);
+
+                size_t I_size_0   = idims[0];
+                size_t I_size_1   = idims[1];
+                size_t T_size_0   = tdims[0];
+                size_t T_size_1   = tdims[1];
+                size_t I_stride_0 = istrides[0];
+                size_t I_stride_1 = istrides[1];
+                size_t T_stride_0 = tstrides[0];
+                size_t T_stride_1 = tstrides[1];
+                kernel(params.i,
```

Comment:
Please measure performance again by referring to https://github.com/moreh-dev/MIOpen/pull/22#discussion_r1621748562.

Reply:
resolved in commit [301a1c6](https://github.com/moreh-dev/MIOpen/pull/25/commits/301a1c66f3bdec977c1d2fdf8203e3451c62bb9c)